{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Chromosome 2q11.1 - analyse 20k segment\n",
    "\n",
    "Analyse a 20 kb segment as 10 x 2 kb intervals.\n",
    "\n",
    "See Sabeti et al. *Positive Natural Selection in the Human Lineage* p.1619 and Altshuler and Donnelly 2005 *A haplotype map of the human genome* Supplementary Table 4 (96.25 - 96.75).\n",
    "\n",
    "Note that the HapMap project cited uses reference human genome assembly version 34.3 (Supplementary information referenced above), which is equivalent to UCSC version hg16 (https://en.wikipedia.org/wiki/Reference_genome#Human_reference_genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pickle, gzip\n",
    "from selectiontest import selectiontest\n",
    "import pysam\n",
    "from vcf import Reader        # https://pypi.org/project/PyVCF/\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "#Include pathname to repository\n",
    "projdir = \"/Users/helmutsimon/repos/NeutralityTest\"\n",
    "if not os.getcwd() == projdir:\n",
    "    os.chdir(projdir)\n",
    "import vcf_1KG\n",
    "\n",
    "\n",
    "path = \"/Users/helmutsimon/\"\n",
    "if not os.getcwd() == path:\n",
    "    os.chdir(path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Examine selection by location along genome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Segment start            = 96985244\n",
      "Tajimas D                = -1.7949661869444085\n",
      "ρ                        = 1.445238911468624\n",
      "Segment start            = 96987244\n",
      "Tajimas D                = -1.9153654950603742\n",
      "ρ                        = 2.9974925525195983\n",
      "Segment start            = 96989244\n",
      "Tajimas D                = -1.7130119378194018\n",
      "ρ                        = 1.8462139199512948\n",
      "Segment start            = 96991244\n",
      "Tajimas D                = -1.8593750761567915\n",
      "ρ                        = 0.9480941482975869\n",
      "Segment start            = 96993244\n",
      "Tajimas D                = -1.6296754226830776\n",
      "ρ                        = 1.911886467329321\n",
      "Segment start            = 96995244\n",
      "Tajimas D                = -1.700811445853704\n",
      "ρ                        = 2.538884001079249\n",
      "Segment start            = 96997244\n",
      "Tajimas D                = -1.6169913089647274\n",
      "ρ                        = -0.13847295405748028\n",
      "Segment start            = 96999244\n",
      "Tajimas D                = -1.907625016303753\n",
      "ρ                        = 1.2080602452924447\n",
      "Segment start            = 97001244\n",
      "Tajimas D                = -1.7179376248028657\n",
      "ρ                        = 0.1139421529243485\n",
      "Segment start            = 97003244\n",
      "Tajimas D                = -1.8127105411463336\n",
      "ρ                        = 3.1088378176266342\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>segstart</th>\n",
       "      <th>tajd</th>\n",
       "      <th>rho</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>96985244</td>\n",
       "      <td>-1.794966</td>\n",
       "      <td>1.445239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>96987244</td>\n",
       "      <td>-1.915365</td>\n",
       "      <td>2.997493</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>96989244</td>\n",
       "      <td>-1.713012</td>\n",
       "      <td>1.846214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>96991244</td>\n",
       "      <td>-1.859375</td>\n",
       "      <td>0.948094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>96993244</td>\n",
       "      <td>-1.629675</td>\n",
       "      <td>1.911886</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>96995244</td>\n",
       "      <td>-1.700811</td>\n",
       "      <td>2.538884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>96997244</td>\n",
       "      <td>-1.616991</td>\n",
       "      <td>-0.138473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>96999244</td>\n",
       "      <td>-1.907625</td>\n",
       "      <td>1.208060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>97001244</td>\n",
       "      <td>-1.717938</td>\n",
       "      <td>0.113942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>97003244</td>\n",
       "      <td>-1.812711</td>\n",
       "      <td>3.108838</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   segstart      tajd       rho\n",
       "0  96985244 -1.794966  1.445239\n",
       "1  96987244 -1.915365  2.997493\n",
       "2  96989244 -1.713012  1.846214\n",
       "3  96991244 -1.859375  0.948094\n",
       "4  96993244 -1.629675  1.911886\n",
       "5  96995244 -1.700811  2.538884\n",
       "6  96997244 -1.616991 -0.138473\n",
       "7  96999244 -1.907625  1.208060\n",
       "8  97001244 -1.717938  0.113942\n",
       "9  97003244 -1.812711  3.108838"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chrom = 2\n",
    "start_hg19 = 96985244\n",
    "end_hg19   = 97005244\n",
    "interval = 2000\n",
    "reps = 10000\n",
    "fname = 'Data sets/1KG variants full/integrated_call_samples_v3.20130502.ALL.panel'\n",
    "panel_all = pd.read_csv(fname, sep=None, engine='python', skipinitialspace=True, index_col=0)\n",
    "panel = panel_all[panel_all['super_pop'].isin(['AFR'])]   \n",
    "vcf_filename = 'Data sets/1KG variants full/ALL.chr' + str(chrom) \\\n",
    "                + '.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz'\n",
    "vcf_file = Reader(filename=vcf_filename, compressed=True, encoding='utf-8')\n",
    "rows = list()\n",
    "num_segs = int((end_hg19 - start_hg19) / interval)\n",
    "for segment in range(num_segs):\n",
    "    seg_start = start_hg19 + segment * interval\n",
    "    print('Segment start            =', seg_start)\n",
    "    seg_end = seg_start + interval\n",
    "    sfs, n, non_seg_snps = vcf_1KG.get_sfs(vcf_file, panel, chrom, seg_start, seg_end)\n",
    "        \n",
    "    tajd = selectiontest.calculate_D(sfs)\n",
    "    print('Tajimas D                =', tajd)\n",
    "    rho = selectiontest.test_neutrality(sfs)\n",
    "    print('\\u03C1                        =', rho)\n",
    "    if len(non_seg_snps) > 0:\n",
    "        print(non_seg_snps)\n",
    "    row = [seg_start, tajd, rho]\n",
    "    rows.append(row)\n",
    "results = pd.DataFrame(rows, columns=['segstart', 'tajd', 'rho'])\n",
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "results.to_csv('Google Drive/Genetics/Bayes SFS/Neutrality test/ch2q11_2kinterval.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
